[BEAM-7520] Fix timer firing order in DirectRunner

je-ik opened the pull request:
Fixes [BEAM-7520].

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

je-ik commented:
Run Direct ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run SQL PostCommit

je-ik commented:
R: @kennknowles @iemejia @robertwb This PR is related to previous [BEAM-7785], but solves the original issue [BEAM-7520] more in a more complex way. I'd be happy if you had time to lokk into this. Many thanks!

je-ik commented:
Run SQL PostCommit

je-ik commented:
Run Direct ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

tvalentyn commented:
Just a heads-up @kennknowles is out of office for 1 week.

je-ik commented:
@tvalentyn I know, thanks. This can wait.

je-ik commented:
@kennknowles would you have time to look into this, please?

reuvenlax commented:
R: @dpmills

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkCallbackExecutor.java:
```diff
@@ -116,14 +119,30 @@ public void callOnWindowExpiration(
+        toFire.add(callbackQueue.poll().getCallback());
+      }
+      if (!toFire.isEmpty()) {
+        CountDownLatch latch = new CountDownLatch(toFire.size());
```
Instead of managing the latch yourself, you could gather a list of futures and wait on it. It would be a bit higher level programming style, and not depend on capturing the latch in the closure, etc.

je-ik replied:
If I understand it correctly, then this in unfortunately currently not possible, because the `executor` is not `ExecutorService`, but simply `Executor`, which returns `void`, instead of `Future`. I cannot simply change it to `ExecutorService`, because `EvaluationContext` passes there `MoreExecutors.directExecutor()`, which is not `ExecutorService`.

je-ik replied:
Hm, my previous comment might suggest, that the change is not needed at all. And maybe it is true. :-) I added this code as part of my "I'm really desperate" period while solving the issue. Just for being sure, that this doesn't break in the future, I'd like to keep it as is, if it is okay.

kennknowles replied:
Fair enough :-). Since it passes and solves a problem, let's definitely merge.

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java:
```diff
@@ -244,27 +249,40 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
         delegateEvaluator.processElement(windowedValue);
       }
 
+      Instant lastFired = null;
```
`@Nullable`

je-ik replied:
:+1:

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java:
```diff
@@ -244,27 +249,40 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
+          lastFired = timer.getTimestamp();
+          WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
+          BoundedWindow timerWindow = windowNamespace.getWindow();
+          delegateEvaluator.onTimer(timer, timerWindow);
```
I'm reading this really quickly, but I think it works only for loops of the same timer, not the general case. And it also leaves the "should fire" implicit only by the fact that it already fired, instead of checking it. I think if you have a unit test where two timer callbacks set each other to "already firing" times you will uncover some issues.

I think there's a simple logic you might have:

1. Add the fired timers in the incoming bundle to a priority queue
2. While any timers to fire, call the first callback
    a. For any timer set by the `@OnTimer` callback, you check if it should fire. If yes, then add to fired timers priority queue. If no, then add it to the committed result.

je-ik replied:
I will add the test and check if there are any issues. Regarding the logic you described - generally, yes, that would be the ideal approach. It would also solve some other issues, that still remain, even after this PR - but - it would require a larger refactoring, because there is currently no simple (or simple and apparent to me) way of knowing if `@OnTimer` method actually setup any new timer. The `delegateEvaluator` would have to signal that, and that boils down to `PushbackSideInputDoFnRunner.onTimer` would have to return the new timers. That is API change in core and I wanted to avoid that. But maybe I've overlooked  some other way.

je-ik replied:
I have a failing test case for the case you described. Will work on that. Many thanks!

je-ik replied:
@kennknowles Fixed the test. I'm not sure if it is exactly what you meant (although the new test really uncovered some flaws in the original implementation). Can you have a look [here](https://github.com/apache/beam/pull/9190/files#diff-37ebd0b68ae7d5b951cfca73aba52cc1R689), please?

kennknowles replied:
I don't think it is quite what I might. Here:

t1.set(50)
t2.set(60)

watermark advances to 100

t1 fires and sets t2 to 40 -- since it is moved earlier it will fire in the same bundle
t2 fires and sets t1 to 30 -- will not fire in the same bundle, but will be committed in the result?

I guess this might get the right result... I was just expecting all the timers that could fire to do it in the same bundle. It is a bit weird that the normal case of timers just coming in order and not setting anything will not fire in one bundle. Did I read that code wrong? This is why I suggested a prio queue, to remove all the logic about `lastFired` and commits. You just fire all the timers, in order (including newly set timers), until the bundle is done. I wonder if there is a use case for always setting a timer for the same time and hoping the bundle will still commit over and over again.

je-ik replied:
Is it possible (and valid) to set timer for time that is *before* the time another timer was set for? That seems weird and actually can cause exception to be thrown in some cases.

je-ik commented:
Run Direct ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Direct ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
@kennknowles I added @ValidatesRunner for timer ordering when using `Create` transform instead of `TestStream` (there was a suspicion that DirectRunner handles it somehow wrong, but it seems to pass), but that apparently broke ValidatesRunner of Dataflow. I'm not sure how to proceed, can you give me a hand?

je-ik commented:
I need to push this forward, because it is prerequisite for more PRs that I have (mostly) ready. But these need to be reliably testable, so sorry if I'm a little impatient. :-)

@kennknowles, if you don't have time for this, would you please delegate this?
Open questions:
 * is the current approach OK? It is tested in terms of not causing deadlocks, performance should be somewhat comparable to current master - I'm aware that the completely correct solution would be to evaluate and add timers to *current* bundle, not rescheduling it for later, but that seems to be major refactor (please, anyone confirm or disprove this statement as it might be wrong)
 * there was observed strange behavior related to `Create` vs. `TestStream` in DirectRunner, but it is not reproducible in this PR. It was observed and described in https://github.com/rezarokni/beam/pull/2#issuecomment-522226493
 * added test for `Create` instead of `TestStream` broke only ValidatesRunner of Dataflow (all other runners seem to pass)

kennknowles commented:
@reuvenlax who is listed here, is also a good reviewer

Sorry! I missed this because I had a mention but failed to add it to my pending review queue. I'll get to it tomorrow.

je-ik commented:
Many thanks! :+1:

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/CommittedResult.java:
```diff
@@ -18,10 +18,10 @@
 package org.apache.beam.runners.direct;
 
 import com.google.auto.value.AutoValue;
+import java.util.Optional;
```
One reason not to use this is that it is not `Serializable` and the Java committee has decided it never should be. I obviously disagree, but until then I don't think `java.util.Optional` should be used.

je-ik replied:
There was a very healthy discussion about usage of `Optional` on flink mailing list: https://lists.apache.org/thread.html/f5f8ce92f94c9be6774340fbd7ae5e4afe07386b6765ad3cfb13aec0@%3Cdev.flink.apache.org%3E
General consensus is that `Optional` should not be used for fields, but only for return values of performance non-critical code. I very much agree with that, if `Optional` is not used as field, its serializability doesn't matter. Moreover, DirectRunner does not rely that heavily on serialization (more than validating, that the user code is serializable). Using standard instead of guava seems to be beneficial to me, without any negative impact.

kennknowles replied:
Yes, that is exactly the logic I am familiar with. We've also had healthy discussions in Beam and seemingly every project. I can't escape null! I share the "null was a mistake, never really made sense in the first place, and should be eliminated from languages" perspective. I believe basic good programming practice to program as though it does not actually exist, except when necessary for compatibility or performance. Using `@Nullable` and `@NonNull` with an extended type checker is a compromise when necessary.

I have not seen any actual evidence for the claim that it causes performance problems. I would easily believe it - It is true that Java is exceptionally inefficient w.r.t. trivial wrappers unless they are well-optimized. But on the thread you linked the only person who brought up real performance said that they could not detect any difference except when _serializing_ it in performance-sensitive paths. So perhaps Java isn't so inefficient at optimizing it away after all.

In any event, please make it a separate change. You can do a bulk change in the codebase if you like. At this point, I am just so tired of this issue which humans are still stuck on that I will let most anything through :-/

je-ik replied:
Yes, I also think that the performance impact of creating one short lived instance should be negligible. I actually tried to measure it, but I didn't get any statistically significant difference in this case. Although, I have seen significant impact in other cases (getting rid of creating useless instance in one place of spark runner gave me about 20% performance gain, but that could have been very specific case). But that is probably not that significant to this PR. :-)

To conclude this, I will drop the commit where I changed the Optional in DirectRunner and start a discuss thread related to making the usage consistent across the code base. Currently there is approx 3:2 usage of java.util vs. guava.

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/CopyOnAccessInMemoryStateInternals.java:
```diff
@@ -22,6 +22,7 @@
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Map;
+import java.util.Optional;
```
ditto

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/EvaluationContext.java:
```diff
@@ -23,6 +23,7 @@
 import java.util.EnumSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
```
ditto

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java:
```diff
@@ -20,6 +20,7 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Map;
+import java.util.Optional;
```
ditto

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java:
```diff
@@ -22,6 +22,7 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Optional;
```
ditto

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/SideInputContainer.java:
```diff
@@ -24,6 +24,7 @@
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
```
ditto

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java:
```diff
@@ -244,27 +250,40 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
+            lastFired,
+            timer.getTimestamp());
+        if (lastFired != null && lastFired.isBefore(timer.getTimestamp())) {
+          pushedBackTimers.add(timer);
```
Now that I look more closely, doesn't this cause the timers to fire one at a time if they are in order, with all the rest pushed back into the commit result? And only when a timer is moved earlier you go to the other branch and fire multiple timers in the same bundle.

je-ik replied:
Generally yes. The problem is that you cannot fire multiple timers at once, because the outcome of firing a single timer is unknown. It might or might not setup some other timer. If it was to be solved, it seems to me it would require `PushbackSideInputDoFnRunner` (and most likely `DoFnRunner` to return set timers, instead of `void`. The other option would be somehow hack into `WatermarkManager` and be able to retrieve "most recent timers set", os something like that (which seems to be fragile and hackish). The correct solution seems to be a too big refactor, given that the performance penalty for the solution proposed in this PR seems to be low.

je-ik commented:
@kennknowles I removed the Optional stuff from this PR, so that it can be focused. I have refactored the code in `StatefulParDoEvaluator` a bit, it now enables firing multiple timers in one bundle (as long as firing any timer does not set or delete any timer that is in the current bundle). If a timer sets timer for time belonging in the current bundle, then all other (not yet fired) timers are pushed back to later bundle. Reason for this is that the logic of handling TimerUpdates lies on `WatermarkManager` and should not be performed by `StatefulParDoEvaluator`. Thanks for the discussion, I didn't dig deeper into the code to see, that I can get quite cleanly instance of `DirectTimerInternals` to get the information that was needed. I think, that now the impact on performance will be truly negligible, because overwhelming majority of cases will fire all timers in bundle that is being processed and only when needed, timers will be pushed back.

je-ik commented:
Run Java PreCommit

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
@kennknowles @reuvenlax The added @ValidatesRunner fails on Dataflow. It looks to me, that it should pass, but cannot tell for sure. Can you check this?

je-ik commented:
For clarity, it is [this one](https://github.com/apache/beam/pull/9190/files#diff-68ed88cc25171e30985dda4534e68c9bR3184)

je-ik commented:
@robertwb Would you have time to look into this please?

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
The problem with Dataflow validatesRunner was apparently only in timeout (strange though, that the same timeout consistently passed when using `TestStream`, but consistently failed with `Create`). Anyway, I removed the timeout and the test passes.

iemejia commented:
This PR seems to be almost there. Anything important still missing @kennknowles?

kennknowles commented:
I haven't had a chance to look again - thanks for the ping.

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java:
```diff
@@ -254,17 +268,23 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
         WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
         BoundedWindow timerWindow = windowNamespace.getWindow();
         delegateEvaluator.onTimer(timer, timerWindow);
+        if (timerInternals.containsUpdateForTimeBefore(currentInputWatermark)) {
```
So if I understand this correctly, the `PriorityQueue` causes the timers in the incoming bundle to be fired in timestamp order. That's good. And then if any timer is set that would already be eligible for firing, it ends processing and pushes the rest back. I think this is not optimal, because you could add timer updates to the priority queue and keep firing any eligible timer. The bundle only needs to end when there are no more eligible timers. But on the other hand, I do think it fixes the problem, so happy to merge it.

je-ik replied:
The problem here is that I don't want to split the logic. `WatermarkManager` is responsible for handling all timer setting/firing/removing, so I didn't want to intrude into the logic from `StatefulParDoEvaluator`. If the current logic is fine, I'd merge it as it is.

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java:
```diff
@@ -829,6 +864,14 @@ public Instant get() {
   private final Set<ExecutableT> pendingRefreshes;
 
+  /**
+   * A set of executables with currently extracted timers, that are to be processed. Note that, due
```
Consistency should only be per (key, window) pair.

je-ik replied:
That is correct. The problem is that `WatermarkManager` works on level of `PTransform`s. I can push the logic down to `TransformWatermarks`, which is the first level, that is somewhat aware of keys (in `extractFiredTimers`), but that would mean, that timers extracted from `AppliedPTransformInputWatermark` would have to be setup again. That is because the logic requires that extracted timer to be fired, or setup again (in the way that `StatefulParDoEvaluator` pushes them back). I think that this makes the logic far too much complicated given that what we gain is the ability to extract and execute timers for various keys of the same PTransform in parallel. I must emphasize, that this is related only to DirectRunner, which is already pretty slow, because it does many validations and its main purpose is testing. And moreover, I'm not convinced that implementing this optimization in reality would bring and significant performance boost (speculative claim, I didn't do any performance testing). Maybe we can merge it as it is and solve performance issues after if they appear (avoid premature optimization with significant impact on readability and understandability of already somewhat complicated code)?

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java:
```diff
@@ -1537,6 +1680,21 @@ public boolean equals(Object other) {
+
+    @Override
+    public String toString() {
+      return "TimerUpdate(key="
```
`MoreObject.toStringHelper` is a great utility for this sort of thing.

je-ik replied:
:+1:

kennknowles commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java:
```diff
@@ -616,6 +649,186 @@ public void processElement(ProcessContext c) {
+   * one triggers data being fed into the second one.
+   */
+  @Test(timeout = 10000)
+  public void testTwoPOutputsInPipelineWithCascade() throws InterruptedException {
```
This seems like a `@Category(ValidatesRunner.class)` test for use in either `CombineTest` or `GroupByKeyTest`.

je-ik replied:
I'm afraid that could cause trouble. This test uses a (little hackish) class `StaticQueue`, that will not work well outside single JVM. That probably will not be problem for many runners, that use a local version of distributed runner for tests (spark, flink), but might definitely cause troubles for fully distributed runners (dataflow). Because the tooling in the test works only in local mode, I'd prefer to keep it for DirectRunner only.

kennknowles replied:
Got it. Makes sense.

kennknowles commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java:
```diff
@@ -616,6 +649,186 @@ public void processElement(ProcessContext c) {
+    Pipeline pipeline = getPipeline();
+    Instant now = new Instant(1500000000000L);
+    Instant end = now.plus(100);
+    PCollection<String> result = pipeline.apply(new TwoTimerTest(now, end));
```
Reading this casually, `TwoTimerTest` is not really clear what is being tested. The code that creates the output and the code that asserts about the expectations are very far from each other. Ideally, you would move the assertion into `TwoTimerTest` somehow. Since `TwoTimerTest` owns the logic, it also owns the expected output.

kennknowles commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java:
```diff
@@ -616,6 +649,186 @@ public void processElement(ProcessContext c) {
+  }
+
+  @Test(timeout = 5000)
+  public void testTwoTimersSettingEachOther() {
```
I believe this is a `@Category(ValidatesRunner.class)` test for `ParDoTest`. And same for all state & timers feature tests.

je-ik replied:
Moved to `ParDoTest`.

kennknowles commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/MultiStepCombineTest.java:
```diff
@@ -64,7 +64,7 @@
       KvCoder.of(StringUtf8Coder.of(), VarLongCoder.of());
 
-  @Test
+  @Test(timeout = 10000)
```
This is unrelated. It may shrink the code review latency to try to separate little changes and just get them merged.

je-ik replied:
Fixed.

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Retest this please

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Samza ValidatesRunner

je-ik commented:
Retest this please

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Spark ValidatesRunner

je-ik commented:
Run Samza ValidatesRunner

je-ik commented:
@kennknowles Unfortunately, samza (at least) doesn't pass the added @ValidatesRunner tests. What would be the suggested procedure? I can create JIRA to address this, but I'm not sure how to make the test @Ignore in samza. The only option I see would be to add a new @Category and ignore that category. Is there a better way?

je-ik commented:
Run Jet ValidatesRunner

je-ik commented:
Run SQL PostCommit

kennknowles commented:
You can exclude the tests by adding a category and tweaking this list: https://github.com/apache/beam/blob/master/runners/samza/build.gradle#L80

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
Run Flink ValidaatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Java_Examples_Dataflow PreCommit

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Samza ValidatesRunner

je-ik commented:
Run Samza ValidatesRunner

je-ik commented:
Run Flink ValidatesRunner

je-ik commented:
Run Java Flink PortableValidatesRunner Streaming

je-ik commented:
@kennknowles this seems to be ready from my side, tests passing. WDYT?

kennknowles commented:
Looks like you need one more "ignore" for the Flink portable ValidatesRunner and Jira for the bug to that runner.

je-ik commented:
Oops, this looks like an endless story. :-)

je-ik commented:
Run Java Flink PortableValidatesRunner Streaming

je-ik commented:
Run Dataflow ValidatesRunner

je-ik commented:
@kennknowles sooo, now it is good, hupefully :crossed_fingers:

kennknowles commented:
My reading of the commit history is that there are many commits which are not meaningful by themselves. There may be some commits that stand alone. But I will squash these commits when I merge unless you would like to do your own custom squashing into some meaningful small set of commits.

je-ik commented:
I'll definitely squash the history to some compact isolated commits before merging. Thanks for the approval!

je-ik commented:
I also plan to create the follow-up JIRAs after the merge and notify dev list.

kennknowles commented:
Missed one: https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/

je-ik commented:
Yikes! Thanks, solution in #9869

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:
```diff
@@ -3488,6 +3499,158 @@ public void onTimer(OutputReceiver<String> r) {
+        elements.add(TimestampedValue.of(KV.of("dummy", "" + i), now.plus(i)));
+      }
+
+      testEventTimeTimerOrderingWithInputPTransform(
```
The elements in the PCollection produced by the Create transform are considered unordered so the stateful DoFn could see (dummy, 1) before it sees (dummy, 0). Doesn't the DoFn need to be marked with `@RequiresTimeSortedInput` in this case?

je-ik replied:
The purpose of the test is to verify, that timers are fired in timestamp order, not that elements arrive ordered. The test already mimics some of the functionality of `@RequiresTimeSortedInput`, which should solve the issue you mention. This test should be a requirement (`UsesStrictTimerOrdering`) for a runer to be able to support `@RequiresTimeSortedInput`. Have you seen any issues with the test?
